app_rstats/main.py

```diff
@@ -753,3 +784,255 @@ def _safe_window_for_geom(dataset, g):
+
+        logging.debug("Computing Y window on its own grid")
+
+        # geom is currently in X CRS (you reprojected it above); convert it to Y CRS
+        if dsx.crs.to_string() != dsy.crs.to_string():
+            _t_xy = Transformer.from_crs(dsx.crs, dsy.crs, always_xy=True)
+            geom_y = shp_transform(
+                lambda x, y, z=None: _t_xy.transform(x, y), geom
+            )
+        else:
+            geom_y = geom
+
+        # Build a tight window on Y raster
+        def _safe_window_for_geom_y(dataset, g):
+            b = g.bounds
+            w = rasterio.windows.from_bounds(*b, transform=dataset.transform)
+            w = w.round_offsets().round_lengths()
+            w = w.intersection(Window(0, 0, dataset.width, dataset.height))
+            if int(w.width) > 0 and int(w.height) > 0:
+                return w
+            xres, yres = map(abs, dataset.res)
+            bpad = (
+                b[0] - 0.5 * xres,
+                b[1] - 0.5 * yres,
+                b[2] + 0.5 * xres,
+                b[3] + 0.5 * yres,
+            )
+            w = rasterio.windows.from_bounds(*bpad, transform=dataset.transform)
+            w = w.round_offsets().round_lengths()
+            w = w.intersection(Window(0, 0, dataset.width, dataset.height))
+            if int(w.width) == 0 or int(w.height) == 0:
+                cx, cy = g.centroid.x, g.centroid.y
+                rr, cc = rasterio.transform.rowcol(dataset.transform, cx, cy)
+                rr = min(max(rr, 0), dataset.height - 1)
+                cc = min(max(cc, 0), dataset.width - 1)
+                w = Window(cc, rr, 1, 1)
+            return w
+
+        win_y = _safe_window_for_geom_y(dsy, geom_y)
+        if win_y is None:
+            raise HTTPException(
+                status_code=400,
+                detail="Empty read window (geometry outside raster_y).",
+            )
+
+        logging.debug("Reading Y window")
+        src_y = dsy.read(1, window=win_y, masked=False).astype(
+            "float64", copy=False
+        )
+        transform_y = dsy.window_transform(win_y)
+
+        logging.debug("Reprojecting Y window into X window grid")
+        dest_y = np.full(arr_x.shape, np.nan, dtype="float64")
+        reproject(
+            source=src_y,
+            destination=dest_y,
+            src_transform=transform_y,
+            src_crs=dsy.crs,
+            src_nodata=nodata_y if nodata_y is not None else None,
+            dst_transform=transform_x,  # X window transform you computed above
+            dst_crs=dsx.crs,
+            dst_nodata=np.nan,
+            resampling=Resampling.nearest,
+            num_threads=0,
+        )
+
+        logging.debug("Applying same polygon mask to Y array")
+        arr_y = np.where(mask, dest_y, np.nan)
+
+        logging.debug("Extracting finite paired values")
+        finite_mask = np.isfinite(arr_x) & np.isfinite(arr_y)
+        x_vals = arr_x[finite_mask]
+        y_vals = arr_y[finite_mask]
```

Comment:
check how the values are paired, like what if different cell sizes etc

Reply:
checked and this part is okay but made a tighter bound on y fixed in d9a856f34a7c9f59de973af61691e8b5f3cb1de1